### PR TITLE
Core BPF impl: Config

### LINKED ISSUE #2: PR the initial Core BPF implementation for review
#### Problem

The Config program has yet to be implemented in Core BPF.

#### Solution

Submit the code for initial review!

### PR DESCRIPTION
**Pull request by buffalojoec**
This PR adds the Core BPF implementation of the Config program.

I think in general it makes sense to reduce dependency on the config program crate in the monorepo, so I've omitted the following for now:

- `create_config_account` (helper)
- `impl ConfigState for StakeConfig`

If that makes sense I'll look into refactoring over there.

Closes #2

### CODE REVIEW

**Comment by kevinheavey** [comment]
Re leaving `create_config_account` in the original config program crate, will that mean keeping the solana-config-program dependency in solana-account-decoder [this discussion](https://github.com/anza-xyz/agave/pull/265) or is there some other plan for that?

**Comment by buffalojoec** [comment]
> Re leaving `create_config_account` in the original config program crate, will that mean keeping the solana-config-program dependency in solana-account-decoder ([this discussion](https://github.com/anza-xyz/agave/pull/265)) or is there some other plan for that?

Actually I was looking at refactoring config program out of the monorepo altogether, if possible. Wherever not possible, I think we can generate a client here and use just the client to supply `account-decoder` with the handlers it needs.

I think what might be best is to just move `get_config_data` and `create_config_account` into the client only, and not in the program anywhere. Wdyt?

**Comment by buffalojoec** [comment]
> I don't think Kinobi generates any of what account-decoder needs for the config program

You don't need to only rely on generated code for a generated client. There's code you can put client-side only.

**Comment by kevinheavey** [comment]
Oh sure I just meant that the generated code isn't anything to do with what account-decoder needs

File: program/tests/functional.rs [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+    let pubkey = Pubkey::new_unique();
+    let signer0 = Keypair::new();
+    let keys = vec![
+        (pubkey, false),
+        (signer0.pubkey(), true),
+        (signer0.pubkey(), true), // Duplicate signer0.
+    ];
+    let my_config = MyConfig::new(42);
+
+    create_config_account(&mut context, &config_keypair, keys.clone()).await;
```
Comment (joncinque) [inline program/tests/functional.rs:480]: I got so confused about this worked at all considering it was trying to initialize with dupes, without realizing that they aren't passed into the program at all :sweat:

File: program/tests/functional.rs [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+        (signer0.pubkey(), true), // Duplicate signer0.
+    ];
+    let instruction =
+        config_instruction::store(&config_keypair.pubkey(), false, dupe_keys, &new_config);
+    let err = context
+        .banks_client
+        .process_transaction(Transaction::new_signed_with_payer(
+            &[instruction],
+            Some(&payer.pubkey()),
+            &[&payer, &signer0], // Config missing from signers.
```
Comment (joncinque) [inline program/tests/functional.rs:547]: It's confusing to read, but the test is passing `config_keypair` as a signer, so let's add it here too.

The old test framework only passes in the `data` portion from the instruction, and not the metas, so although the instruction is created with `false`, it's then passing in the config as a signer at https://github.com/anza-xyz/agave/blob/85c14b4e4f003f2e354500ce7bdf56995f6eb4c7/programs/config/src/config_processor.rs#L691

File: program/tests/functional.rs [hunk 3]
```diff
@@ -0,0 +1,806 @@
...
+        .unwrap();
+    assert_eq!(
+        err,
+        TransactionError::InstructionError(0, InstructionError::MissingRequiredSignature)
+    );
+}
+
+// [Core BPF]: I don't think this test is necessary.
+// This test is really just making sure the runtime doesn't panic when no
+// accounts are provided. Not relevant to BPF programs, really.
```
Comment (joncinque) [inline program/tests/functional.rs:645]: Panicking isn't great since the runtime needs to handle it, so let's keep the test

File: program/tests/functional.rs [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+        .unwrap_err()
+        .unwrap();
+    assert_eq!(
+        err,
+        TransactionError::InstructionError(0, InstructionError::InvalidAccountOwner)
+    );
+}
+
+#[tokio::test]
+async fn test_maximum_keys_input() {
```
Comment (joncinque) [inline program/tests/functional.rs:728]: Love this new test!

File: program/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "scbpf-config"
+version = "0.1.0"
+edition = "2021"
+readme = "./README.md"
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/Cargo.toml:5]: I just copied this bit from the lookup table one, heh.

[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/Cargo.toml:5]: I would actually say no, though. That README may have documentation on the program and the clients, as well as maybe local development setup tips. The program README should be more specifically geared toward the program crate itself. I'll just add a dummy README to the program for now.

File: program/tests/functional.rs (hunk 3 above)
Comment (buffalojoec) [reply program/tests/functional.rs:645]: Ohh I was alluding to the fact that it's a bit redundant considering it's now a BPF program, and the program-runtime should be more than tested to ensure zero-account inputs don't crash the runtime.

File: program/tests/functional.rs [hunk 6]
```diff
@@ -0,0 +1,806 @@
...
+}
+
+#[tokio::test]
+async fn test_process_store_without_config_signer() {
+    // [Core BPF]: This test actually has nothing to do with signatures.
+    // Instead, it's testing what happens when the config account is not
+    // included in the accounts list. In the original builtin test, it's
+    // marked as a signer in the instruction, but since that won't pass
+    // transaction sanitation from the client here, this version of that
+    // same tests looks even more strange.
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/tests/functional.rs:264]: Sorry, do you mean just adjust the naming and scope of the test? To reflect what it's actually doing?

File: program/tests/functional.rs [hunk 7]
```diff
@@ -0,0 +1,806 @@
...
+            &[&payer, &signer0], // Missing signer1.
+            context.last_blockhash,
+        ))
+        .await
+        .unwrap_err()
+        .unwrap();
+    assert_eq!(
+        err,
+        TransactionError::InstructionError(0, InstructionError::MissingRequiredSignature)
+    );
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/tests/functional.rs:439]: Sounds good!

File: program/tests/functional.rs (hunk 1 above)
Comment (buffalojoec) [reply program/tests/functional.rs:480]: Hah yeah just allocating there.

Comment (buffalojoec) [reply program/tests/functional.rs:480]: I wanted to leave things as they were, but if we want, we can adjust some things that are weird. For example, the instruction builder could just accept a `num_keys` and we could change `ConfigKeys::serialized_size` to use dummy keys to calculate length.

File: program/tests/functional.rs (hunk 2 above)
Comment (buffalojoec) [reply program/tests/functional.rs:547]: You're right I missed that bit! Added.

File: program/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (joncinque) [reply program/Cargo.toml:5]: That works too!

File: program/tests/functional.rs (hunk 1 above)
Comment (joncinque) [reply program/tests/functional.rs:480]: Nah no worries for this round of changes, I just got *super* confused

File: program/tests/functional.rs (hunk 3 above)
Comment (joncinque) [reply program/tests/functional.rs:645]: I see what you mean. My point is that panicking is still bad for BPF programs since the VM needs to handle it, so this test is definitely still useful. So I would remove the comment

File: program/tests/functional.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (joncinque) [reply program/tests/functional.rs:264]: Ah, no, sorry, I meant that I think the test is supposed to check that flipping the `is_config_signer` flag to false still works, but that it wasn't properly written.

No other test up to this point checks that it works to do:

```
config_instruction::store(&config_keypair.pubkey(), false, keys.clone(), &my_config);
```

Since previously, they've all expected the config account to sign. So maybe this test should show that it works when you set the flag to `false` and *not* sign with the config account. What do you think?

[the remark this answers is not in the document]
Comment (buffalojoec) [reply program/tests/functional.rs:264]: To me, it seemed like the original was checking that even if you set `config_is_signer: false`, if you provide the wrong account at index `0` it will throw `InvalidAccountData`.

Either way, the act of storing keys in the config account with `config_is_signer: false` is actually covered in later tests via "update". The program actually forbids you from doing `config_is_signer: false` the first time you write keys to it. I'm not sure if this is a bug or a feature...

```rust
let current_signer_keys: Vec<Pubkey> = current_data
    .keys
    .iter()
    .filter(|(_, is_signer)| *is_signer)
    .map(|(pubkey, _)| *pubkey)
    .collect();
if current_signer_keys.is_empty() {
    // Config account keypair must be a signer on account initialization,
    // or when no signers specified in Config data
    if !is_config_account_signer {
        return Err(InstructionError::MissingRequiredSignature);
    }
}
```

[the remark this answers is not in the document]
Comment (joncinque) [reply program/tests/functional.rs]: Oh right, that's a good point... eh all right, let's leave it as is then

**Review by joncinque — APPROVED** [review]
One more program conversion for the books, great job!

config
